logstash-core/src/test/java/org/logstash/log/LoggingSpyResource.java

```diff
@@ -0,0 +1,40 @@
+package org.logstash.log;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.core.*;
+import org.apache.logging.log4j.test.appender.ListAppender;
+import org.junit.rules.ExternalResource;
+
+import java.util.List;
+
+public class LoggingSpyResource extends ExternalResource {
+
+    private static final String APPENDER_NAME = "spyAppender";
+
+    private final Logger loggerToSpyOn;
+    private final Level levelSnapshot;
+    private final ListAppender appender = new ListAppender(APPENDER_NAME);
+
+    public LoggingSpyResource(final org.apache.logging.log4j.Logger loggerToSpyOn) {
+        this.loggerToSpyOn = (Logger) loggerToSpyOn;
+        this.levelSnapshot = this.loggerToSpyOn.getLevel();
```

Comment:
Looks like we hit a similar issue to https://github.com/elastic/logstash/pull/16719#discussion_r1856937636 

i like this pattern, i wonder if it would be worth going through the tests in general an making sure we are cleaning up state everywhere. (not in this PR 😅 )

Reply:
I agree. -> https://github.com/elastic/logstash/issues/16896